File: src/main/java/december_2024/MergeTwoLinkedLists.java
package december_2024;

import java.util.List;

public class MergeTwoLinkedLists {

    public static void main(String[] args) {
        ListNode list1 = new ListNode(1);
        list1.next = new ListNode(2);
        list1.next.next = new ListNode(4);

        ListNode list2 = new ListNode(1);
        list2.next = new ListNode(3);
        list2.next.next = new ListNode(4);

        MergeTwoLinkedLists mergeTwoLinkedLists = new MergeTwoLinkedLists();
        ListNode listNode = mergeTwoLinkedLists.mergeTwoLists(list1, list2);
        System.out.println(listNode);
    }

    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if (list1 == null) {
            return list2;
        }
        if (list2 == null) {
            return list1;
        }
        ListNode temp = new ListNode(0);
        if (list1.val <= list2.val) {
            temp.next = list1;
        } else {
            temp.next = list2;
        }
        ListNode p1p = null;
        ListNode p1 = list1;
        ListNode p2 = list2;

        while (p1 != null && p2 != null) {
            if (p1.val <= p2.val) {
                p1p = p1;
                p1 = p1.next;
            } else {
                if (p1p != null) {
                    p1p.next = p2;
                }
                p1p = p2;
                p2 = p2.next;
                p1p.next = p1;
            }
        }
        if (p1 == null) {
            p1p.next = p2;
        }

        return temp.next;
    }

    static class ListNode {
        int val;
        ListNode next;

        public ListNode(int val) {
            this.val = val;
        }
    }

}

File: python/merge_two_linkedlists.py
class Solution(object):
    # O(n) time | O(n) space
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        dummy = ListNode(0, None)
        curr = dummy

        def helper(list1, list2, curr):
            if not list1 and not list2:
                return
            if not list1:
                curr.next = list2
                return
            if not list2:
                curr.next = list1
                return
            if list1.val <= list2.val:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next

            curr = curr.next
            helper(list1, list2, curr)
            
        helper(list1, list2, curr)

        return dummy.next


    # O(n) time | O(1) space
    def mergeTwoListsIter(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        dummy = ListNode(0, None)
        curr = dummy

        while list1 and list2:
            if list1.val <= list2.val:
                curr.next = list1
                list1 = list1.next
            else:
                curr.next = list2
                list2 = list2.next
            curr = curr.next
        
        if list1:
            curr.next = list1
        if list2:
            curr.next = list2

        return dummy.next


class ListNode(object):
    def __init__(self, val, next):
        self.val = val
        self.next = next
    
    def __repr__(self) -> str:
        res = str(self.val)
        if self.next:
            res += str(self.next)
        return res


list1 = ListNode(1, ListNode(2, ListNode(3, None)))
list2 = ListNode(1, ListNode(3, ListNode(4, None)))

solution = Solution()
print(solution.mergeTwoLists(list1, list2))